### ISSUE TITLE
Yank state machine out of Worker class

### ISSUE THREAD

**Issue by crusaderky** [dask/distributed#6476]
This issue is a tracker for the worker state machine refactor, as described in the wider epic #5736.

In scope for this issue is to yank out all state machine logic from Worker and create two new classes, WorkerState and BaseWorker.

Writing additional unit tests, beyond tweaking the already existing ones, is out of scope.

Contributing PRs:
- #6579
Blocked by:
  - #6581 
  - #6566
  Blocked by:
    - #5896
    - #6410 
    - #6475
    Blocked by:
      - #6474 
    - #6512 
    - #6564

**Comment by fjetter** [comment]
apparently _blocks_ also acts as a "please close ticket" keyword 🤔

**Comment by zklaus** [comment]
Just wanted to leave some words of gratitude and encouragement. It's great to see this move forward; memory leaks and deadlocks due to worker transitions have become a real problem in our production code lately, so your work on this is greatly appreciated!

**Comment by crusaderky** [comment]
I'll have to move `_notify_plugins` from Worker to WorkerState and I'll have to share the `Worker.plugins` dict by reference with WorkerState. This is because of
https://github.com/dask/distributed/blob/9b8172b56c749eab621d69e747ed76bf2841d0f2/distributed/worker.py#L2783

@fjetter were you expecting this? Do you have an alternative design in mind?

**Comment by crusaderky** [comment]
I'm going to apply the following changes to methods of Worker. This will potentially break third party code that directly interact with (what we believe is) internal API.
Please review and speak up if you see any issues.

- These attributes will be shared by reference between BaseWorker and WorkerState, like it already happens for `WorkerMemoryManager`:
  - `data`
  - `plugins`
- These methods will move from `Worker` to `WorkerState` and will become private, with no deprecation cycle:
  - `transitions()`
  - `transition_<begin>_<finish>()`
  - `ensure_task_exists()`
  - `validate_task_<state>()`
- These attributes and methods will move from Worker to WorkerState, with an accessor in Worker which will raise a DeprecationWarning:
  - `tasks`
  - `log`
  - `stimulus_log`
  - `story()`
  - `stimulus_story()`
  - `total_out_connections`
  - `comm_threshold_bytes`
  - `data_needed`
  - `data_needed_per_worker`
  - `ready`
  - `constrained`
  - `executing_count`
  - `executed_count`
  - `long_running`
  - `has_what`
  - `in_flight_tasks`
  - `in_flight_workers`
  - `busy_workers`
  - `comm_bytes`
  - `waiting_for_data_count`
  - `generation`
  - `target_message_size`
  - `transition_counter`
  - `transition_counter_max`
  - `available_resources`
  - `_missing_dep_flight`
  - `_executing`
  - `executing_count` property
  - `_in_flight_tasks`
  - `in_flight_tasks` property
  - `validate_task()`

- These methods will be deprecated:
  - `keys()` (use `state.data.keys()` instead)
  - `stateof()` (no replacement)

**Comment by fjetter** [comment]
This list looks good to me.
